android/src/main/java/com/reactnativeone/OneModule.java

```diff
@@ -72,62 +80,102 @@ public void init(String siteKey, String touchpointUri, String apiKey,
+  public void sendInteraction(String interactionPath, ReadableMap propertiesMap, final Promise promise) {
     HashMap<String, String> properties = getPropertiesFromReadableMap(propertiesMap);
     final OneRequest sendInteractionRequest = new OneRequest.Builder()
-      .interactionPath(new OneInteractionPath(URI.create(interaction)))
+      .interactionPath(new OneInteractionPath(URI.create(interactionPath)))
       .properties(properties)
       .build();
-    final OneCall sendInteractionCall = One.sendInteraction(sendInteractionRequest);
-
-    if (promise == null) {
-      sendInteractionCall.enqueue(null);
-    } else {
-      sendInteractionCall.enqueue(new OneCallback() {
-        @Override
-        public void onSuccess(@NotNull OneResponse oneResponse) {
-          notifyResult(promise,
-            "{\"captures: " + oneResponse.getCaptures().toString() + ", " +
-                  "\"interactionPath\": \"" + oneResponse.getInteractionPath().getValue().toString() + "\", " +
-                  "\"optimizations: " + oneResponse.getOptimizations().toString() + ", " +
-                  "\"statusCode\": " + oneResponse.getHttpStatusCode() + ", " +
-                  "\"tid\": \"" + oneResponse.getTid() + "\", " +
-                  "\"trackers\": " + oneResponse.getTrackers().toString() + "}"
-          );
-        }
-
-        @Override
-        public void onError(@NotNull OneSDKError oneSDKError) {
-          notifyProblem(promise, "" + oneSDKError.getSystemCode(), oneSDKError.getErrorMessage());
+    executor.submit(() -> {
+      try {
+        OneResponse response;
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+          response = One.sendInteraction(true, sendInteractionRequest).join();
+        } else {
+          response = One.sendInteractionLegacySupport(true, sendInteractionRequest).join();
         }
-
-        @Override
-        public void onFailure(@NotNull OneAPIError oneAPIError) {
-          notifyProblem(promise, "" + oneAPIError.getHttpStatusCode(), oneAPIError.getErrorMessage());
-        }
-      });
-    }
+        One.processResponse(response);
+        WritableNativeMap responseMap = responseObjectToReadableMap(response);
+        notifyResult(promise, responseMap);
+      } catch (ExecutionException error) {
+        notifyProblem(promise, Integer.toString(error.hashCode()), error.getLocalizedMessage());
+        Log.e(NAME, "[Thunderhead] Send Interaction Completion Error: " + error.getCause());
+      } catch (OneSDKError error) {
+        notifyProblem(promise, Integer.toString(error.getSystemCode()), error.getLocalizedMessage());
+        Log.e(NAME, "[Thunderhead] Send Interaction SDK Error: " + error.getErrorMessage());
+      } catch (OneAPIError error) {
+        notifyProblem(promise, Integer.toString(error.getHttpStatusCode()), error.getLocalizedMessage());
+        Log.e(NAME, "[Thunderhead] Send Interaction Api Error: " + error.getErrorMessage());
+      } catch (Exception error) {
```

Comment:
I don't see `CompletionException` caught to then unwrap and give better logging like we talked about earlier. 

When the underlying native API throws an exception, a `CompletionException` will be thrown which will contain the real error. Please catch that code path and log the httpapi/sdk error.

Reply:
Updated.  It does look a little messier now though because CompletionException is only supported on Android N+ so there are two separate try/catches now